src/Data/Bin/Base.agda

```diff
@@ -0,0 +1,99 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Arithmetic of binary represented natural numbers. Initial part.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Bin.Base where
+
+open import Algebra.FunctionProperties using (Op₂)
+open import Data.Nat using (ℕ) renaming (suc to 1+_; _*_ to _*ₙ_)
+open import Function using (_∘_)
+open import Relation.Binary using (Decidable)
+open import Relation.Binary.PropositionalEquality using (_≡_; refl; cong)
+open import Relation.Nullary using (yes; no)
+
+------------------------------------------------------------------------
+-- Definition
+
+data Bin : Set where
+  0#    : Bin
+  2suc  : Bin → Bin    -- n → 2*(1+n)  arbitrary nonzero even
+  suc2* : Bin → Bin    -- n → 1 + 2*n  arbitrary odd
+
+------------------------------------------------------------------------
+-- Basic operations
+
+2* : Bin → Bin
```

Comment:
Looking through your code, and I think `double` would be a much better name for this function. That a) should remove some of the confusion with the `_*_` operation and b) is a much more similar name to `suc` and `pred`, the other special case functions.

Reply:
> *  Not renaming the ℕ operations when importing them

Fixed.

> *   Renaming the 2* constructor to double (see new comments).

Done.

> *   Using the +-*-solver from Data.Nat.Solver to shorten some of the proofs (see new 
> comments).

Done. See ``solve`` usage in Bin.Properties

> Renaming the Bin constructors to ``zero, 1+[2*_] and 2*[1+_]``

See my question about this.